tnz/zti.py

```diff
@@ -2694,6 +2702,55 @@ def __refresh(self):
 
         self.stdscr.refresh(_win_callback=self.__set_event_fn())
 
+    def __register_macros(self):
+        import importlib.util
+        import sys
+        import types
+
+        if not os.path.isdir(self.macros_dir):
+            _logger.exception(f"{self.macros_dir} is not a directory")
+            return
+
+        for macro_file in os.listdir(self.macros_dir):
+            if not macro_file.endswith(".py"):
+                continue
+
+            if len(macro_file.split('.')) > 1:
+                continue
+
+            macro_name = macro_file.split('.')[0]
+
+            # Ignore macros which already exist
+            if f"do_{macro_name}" in self.get_names():
+                continue
+
+            macro_path = os.path.join(self.macros_dir, macro_file)
+
+            # Import the user macro as a module
+            macro_spec = importlib.util.spec_from_file_location(
+                f"module.{macro_name}", macro_path)
+            macro_module = importlib.util.module_from_spec(macro_spec)
+            sys.modules[f"module.{macro_name}"] = macro_module
+            macro_spec.loader.exec_module(macro_module)
+
+            # Find the function
+            if hasattr(macro_module, f"do_{macro_name}"):
+                # Create a new bound method for the `Zti` class for this
+                # function
+                setattr(Zti, f"do_{macro_name}",
+                        types.MethodType(
+                            getattr(macro_module, f"do_{macro_name}"),
```

Comment:
You may have noticed that many of the `do_<command>` methods call `self.__bg_wait_end()`. That is because a background thread runs while the user is at the prompt and that call shuts down that background thread so that the session(s) can be used in a thread-safe manner. That is something that is definitely needed here. Without it. both the thread running the `do_<command>` method and the background thread could be using tnz methods that are not thread-safe. I suggest you define a method that _wraps_ the macro method in order to accomplish this. See `do_plugin()` as an example.

Reply:
Another thing to model after `do_plugin()` ... 
```python
                    with ati.ati.new_program(share_sessions=True):
                        plugin(arg, **kwargs)
```

That helps ensure that the ati environment provided to the plugin is _clean_ and that any variables set by the plugin do not _leak_ into the current environment. Unless your use case requires non-shared/global ati variables as input to or output from the macro, I suggest being consistent with `do_plugin()`.